nca/FWRules/ConnectivityGraph.py

```diff
@@ -133,26 +129,133 @@ def _creates_cliqued_graph(directed_edges):
+        return directed_edges, bicliques_edges, bicliques_nodes
+
+    @staticmethod
+    def _find_equal_groups(peers_edges):
+        """
+        find set of peers that have the same io edges and the same namespace
+        :param: peers_edges: dict peer -> [peer edges]
+        return: list
+        list: list of lists, each list is a group of peers
+        list: list of peers that are not in any group
+        """
+        groups_dict = defaultdict(list)
+        for peer, peer_edges in peers_edges.items():
+            groups_dict[(peer.namespace, peer_edges)].append(peer)
+        groups = [sorted(peer, key=str) for peer in groups_dict.values() if len(peer) > 1]
+        left_out = [peer[0] for peer in groups_dict.values() if len(peer) == 1]
+        return groups, left_out
+
+    def _get_equals_groups(self):
+        """
+        _get_equals_groups find in the graph sets of peers that has the same connections.
+        it find two kinds of groups:
+        1. set of peers that have the same input and the same output edges
+        2. set of peers that have the same input and the same output edges and also connected to each other
+
+        the algorithm:
+        1. for each peer, collect the list of input and output edges.
+        2. for each peer, add self pointing edge
+        3. find group of peers that share the same edges list (the group of kind (2))
+        4. remove the peers that are already grouped
+        5. remove the self pointing edge
+        6. find group of peers that share the same edges list (the group of kind (1))
+
+        return: list( pair(list, list))
+        list: list of the pairs, each pair is the group and the self edges of the group
+        """
+        # for each peer, we get a list of (peer,conn,direction) that it connected to:
+        peers_edges = {peer: [] for peer in set(self.cluster_info.all_peers)}
+        for connection, peer_pairs in self.connections_to_peers.items():
+            for src_peer, dst_peer in peer_pairs:
+                if src_peer != dst_peer and connection:
+                    peers_edges[src_peer].append((dst_peer, connection, False))
+                    peers_edges[dst_peer].append((src_peer, connection, True))
+
+        # for each peer, adding a self edge only for connection that the peer already have:
+        for peer, peer_edges in peers_edges.items():
+            for connection in set(c[1] for c in peer_edges):
+                peers_edges[peer].append((peer, connection, False))
+                peers_edges[peer].append((peer, connection, True))
+        peers_edges = {peer: frozenset(edges) for peer, edges in peers_edges.items()}
+
+        # find groups of peers that are also connected to each other:
+        connected_groups, left_out = self._find_equal_groups(peers_edges)
+        connected_groups = [(group, set(conn[1] for conn in peers_edges[group[0]])) for group in connected_groups]
```

Comment:
It's better to return just one ConnectionSet object rather than a set of ConnectionSet objects. If needed, it's possible to OR all the ConenctionSets together, although I think there should never be more than one ConnectionSet here. This is how `self.connections_to_peers` is built - a <source peer, destination peer> pair has only one ConnectionSet attached.

Reply:
I do not want to use OR here, it basically will add a new connection, and I wrote my code assuming that self.connections_to_peers is not changing . 
one solution is to return a set of  ConnectionSet objects, and later to create only one self loop, with more than one label.
the second solution is to assume that there is only one connection between two peers. are we defiantly  sure it is that case?

Reply:
I consulted with @adisos, and she confirmed that each <source peer, destination peer> pair has only one ConnectionSet attached in `self.connections_to_peers`.

Reply:
fixed.
also fixed a bug:
instead of taking all the connection of the first peer in the group, I should take only a connection of an edge **inside** the group